wp-resizer fixes and improvements

aleixsuau opened the pull request:
https://community.openproject.com/wp/33120

wp-resizer fixes:
- [x] default min-width and
- [x] avoid text selection while the user is dragging.

wp-resizer improvements: 
- [x] show the resizer in red color when the min-width has been reached.

aleixsuau committed:
Resizer fixes: set default min-width and avoid
text selection while user is dragging.
Resizer improvements: show the resizer in red color when the min-width
 has been reached.

wielinde commented:
@HDinger Could please also review this PR of @aleixsuau? You were the main author of the original code. For me the code looks good and I checked that it works as it should.

HDinger commented on frontend/src/app/components/resizer/wp-resizer.component.ts:
```diff
@@ -136,10 +149,32 @@ export class WpResizerDirective extends UntilDestroyedMixin implements OnInit, A
+      newValue = this.elementMinWidth;
+
+      // Show the resizer red when it reaches its limit (min-width)
+      if (this.resizer.style.color !== 'red') {
```
Please do not use inline styles in .ts/.js code. Instead check for a class and add/remove this class accordingly. 
I don't think that we already have an existing class, so you may need to add a new one `-error` or whatever you want to call it and use one of our existing style variables for the actual color.

aleixsuau replied:
Thank you for your feedback @HDinger 

1 - It is flickering because a set a timeout of 500ms to remove the style, so if you drag longer the style is removed and then it is set again. I realize that there is a 'resized' event, so I'll move there the css cleaning to solve this.

2 - Ok, I'll do that. A few questions regarding this:
- Should I place this new '-error' class in the component? I'm trying to @import from _global.scss but I got "SassError: no mixin named exports". How do you use sass variables on your components? I couldn't find any case.
- Do you have any place where you store classes for global use? Should I place this new '-error' class there?

Thanks and excuse so many questions.

HDinger replied:
You have basically two options for the class: either you add it limited the wp resizer component, or you add a globally available class. In my eyes, a global error styling class makes sense as we do not have one yet. In this case, you can add it to the file `_generic.sass`. There, we have all kind of classes that are not limited to a component and useful on multiple places. 

Regarding the variables: The (global) CSS variables that we have, are defined in the `design.rb`. You can use these variables without any import. The syntax slightly differs between angular encapsulated styles and the styles which are lying under `app/assets/stylesheets/`. For the first one, you write `color: var(--variable-name)`. For the later, you use the normal CSS variable syntax: `color: $variable-name`.

So in this case, the following code should be added to the `_generic.sass`. Afterwards you only have to toggle/check for the class name.

```
.-error
  color: $content-form-error-color
```

Don't worry about asking many questions :) Please, let me know if something is still unclear.

HDinger reviewed (changes requested):
Looks good overall. Just two minor remarks:

1. The color change of the icon flickers when I stop with the mouse (see gif). 
![flickering_resizer](https://user-images.githubusercontent.com/7457313/79986476-63b0e480-84ac-11ea-9713-00c602af7f1a.gif)

2. Please have a look at my remark in the code.

One other thing, that is just my opinion and no need for changes right now: Personally, I am not a fan of the red highlighting. I understand that a visual hint is useful to understand why the view does not shrink any further. However, I don't think that changing the color is really explanatory. Unfortunately, I don't have a better suggestion right now, so let's give it try.

oliverguenther commented on frontend/src/app/components/resizer/wp-resizer.component.ts:
```diff
@@ -136,10 +149,32 @@ export class WpResizerDirective extends UntilDestroyedMixin implements OnInit, A
+        this.resizer.style.color = 'red';
+
+        // Avoid triggering change detection for all the app
+        this.ngZone.runOutsideAngular(() => {
```
Without having looked at this further, can't we make this component `OnPush` and avoid these tricks?

aleixsuau replied:
Hi @oliverguenther ,
Thank you for your feedback.

This common performance optimization utility (runOutsideAngular) is aimed to avoid triggering Change Detection for the entire app, in this case just for updating one component's view. Using OnPush is an improvement (just applied, thanks), but it wouldn't avoid this extra Change Detection caused by the setTimeout as this sample illustrates: https://stackblitz.com/edit/change-detection-onpush-timeout-triggers-app-change-detection?file=src/app/child/child.component.ts.

Maybe this kind of performance optimization is not necessary on this app right now, I made it just in case.

Looking forward to hearing your thoughts on this.

oliverguenther replied:
Hi @aleixsuau, thanks for your explanation. First-off, I missed this in the previous comment. Welcome to the team and glad you're here! :partying_face: 

You're right that this avoids an extra cycle. With `onPush`, that at least gets reduced to this component + tree so it should be negligible.

I'm fine with leaving it in. That being said, we did not use this pattern much so far as it only works for direct DOM changes and in this case, I would have written in differently - e.g., with switching a variable to apply a conditional class. I think this is also relevant to Henriettes remark.

Does this make sense?

HDinger replied:
>  First-off, I missed this in the previous comment. Welcome to the team and glad you're here! 🥳

Oh my good, I am sorry! I missed that, too 🙈 Welcome on board 🥳

aleixsuau replied:
Thank you very much for the welcoming @oliverguenther and @HDinger, I'm really excited about becoming part this project :) 

I've made a few changes that could solve the issues you pointed out.

> with switching a variable to apply a conditional class
@oliverguenther Do you mean using ngClass or similar? Since the element we are styling is placed outside this component, the only way that came to my mind was using the DOM API.

oliverguenther replied:
Thanks for the explanation, I wasn't aware that this styled outside of its component, then this of course is all perfectly reasonable. 👍

aleixsuau replied:
Great, thanks 👍

aleixsuau committed:
min-width reached feedback improvements:
-  Moved styles to the -error glocal class
- Styles clean up moved to the 'resizeEnd' method
(no need for ngZone nor ChangeDetector)

HDinger commented on modules/bim/app/assets/stylesheets/bim/ifc_viewer/generic.sass:
```diff
@@ -1,6 +1,7 @@
 // -------------------------- GENERIC --------------------------
 @import "openproject/variables"
 @import "openproject/mixins"
+@import "openproject/generic"
```
Why did you import the generic file here? It should not be necessary, as the all styles are loaded in the BIM module, too. Thus when applying the `-error` class, it should already style the resizer accordingly.

aleixsuau replied:
Hi @HDinger ,

Thank you for your feedback.

If I remove this line it doesn't work for me. I realized that the variables, mixins, and other styles used in the BIM project were imported here. I imported them and it worked so I thought this was the way to go.

Do you have any idea about why it's not working on my machine without importing the generic partials here?

HDinger replied:
Mhhh that is weird. For comparison, this is how the compiled code looks for me (without importing the generic file):

<img width="1525" alt="Bildschirmfoto 2020-04-28 um 09 02 24" src="https://user-images.githubusercontent.com/7457313/80457247-47d79380-892f-11ea-868e-894c5193d8bf.png">

Could you please check, how it looks for you? I assume, that some of the styles are not correctly loaded, that is why you need to import the generic.sass and still see the resizer in red.

aleixsuau replied:
It seems that importing the generic.scss gives it precedence over the openproject.scss
![Captura de pantalla 2020-04-28 a las 9 28 11](https://user-images.githubusercontent.com/25689432/80460366-0695b280-8934-11ea-985b-75d53ce7b4df.png).
Adding !important and removing the import works, thanks.

HDinger replied:
Ah okay, that makes sense..

HDinger commented on app/assets/stylesheets/openproject/_generic.sass:
```diff
@@ -116,3 +116,6 @@
   z-index: 0 !important
+
+.-error
+  color: $content-form-error-color
```
Unfortunately, the color is overwritten by the resizer itself. Thus, there is no change in color right now. So please, add an `!important` here. Normally we try to avoid this. However, the are exceptions: One is for example a generic style class that shall definitely overwrite the normal element style (like in this case).

aleixsuau replied:
It works on my machine (I see the resizer in red color when it reaches the min-width), so definitely I have to check what is happening on my local. I'll add the !important, thanks.

HDinger commented on frontend/src/app/components/resizer/wp-resizer.component.ts:
```diff
@@ -133,13 +145,35 @@ export class WpResizerDirective extends UntilDestroyedMixin implements OnInit, A
     const event = new Event(this.resizeEvent);
     window.dispatchEvent(event);
+
+    if (this.resizer.classList.contains('-error')) {
```
As you use this code block three times only with slight differences, it may make sense to extract it into a separate method:
```
private toggleErrorClass(val:boolean){
  if (..contains...) {
    jQuery(this.resizer).toggleClass('-error', val)
  }
}
```

aleixsuau replied:
Thanks for your feedback @HDinger , it made me think. I understand what you mean, and I agree with extracting logic to reusable methods.

I think this case is slightly different from a toggle scenario though. In order to avoid unnecessary DOM manipulations (resizeMove is triggered a lot of times), what I wanted here is to add '-error' class only if it isn't present (lines 149 & 173) and remove the '-error' class only if it is present (line 167). So it is a kind of a toggleIf, something like:

```
private toggleErrorClass(val:boolean){
  if ((val && !this.resizer.classList.contains('-error')) || 
      (!val  && this.resizer.classList.contains('-error'))) {
    jQuery(this.resizer).toggleClass('-error', val)
  }
}
```

To me, it looks much clearer the DOM API way, it is standard, more semantic and I don't see too much logic to abstract away in this case. I also try to avoid using third party libraries for DOM manipulations when possible.

Mixing both perspectives, I came up with this:

```
private manageErrorClass(shouldBePresent:boolean) {
    if (shouldBePresent && !this.resizer.classList.contains('-error')) {
       this.resizer.classList.add('-error');
    }

    if (!shouldBePresent && this.resizer.classList.contains('-error')) {
      this.resizer.classList.remove('-error');
    }
}
```

Would it be ok for you?

Looking forward to hearing your thoughts on this.

HDinger replied:
I like your suggestion. 👍  In general, of course, we should try to make as few calls as possible. My suggestion was also not meant to simply adopt it, but to clarify what I mean. For me, it's just important that similar code parts don't duplicate themselves too often. And this is fulfilled with your suggested method. :)

aleixsuau replied:
Cool, thanks 👍

HDinger reviewed (changes requested):
Hi @aleixsuau 

thanks for your changes and explanations. 👍  There are only some minor things left, please have a look at my comments.

HDinger commented:
Could you please also change the target of this PR to the branch `release/10.5`. Thanks a lot :)

aleixsuau committed:
- Removed generic.sass import
- Error class management refactor

HDinger reviewed (approved):
Thanks a lot for your patience! 👍
